src/proxy/pool.rs

```diff
@@ -12,80 +12,398 @@
+    socket_factory: Arc<dyn SocketFactory + Send + Sync>,
+    cert_manager: Arc<SecretManager>,
+    drainer: Watch,
+    // this must be a readlockable list-of-locks, so we can lock per-key, not globally, and avoid holding up all conn attempts
+    established_conn_writelock: Arc<RwLock<HashMap<u64, Option<Mutex<()>>>>>,
 }
 
-impl Pool {
-    pub fn new() -> Pool {
+impl WorkloadHBONEPool {
+    // Creates a new pool instance, which should be owned by a single proxied workload.
+    // The pool will watch the provided drain signal and drain itself when notified.
+    // Callers should then be safe to drop() the pool instance.
+    pub fn new(
+        cfg: crate::config::Config,
+        socket_factory: Arc<dyn SocketFactory + Send + Sync>,
+        cert_manager: Arc<SecretManager>,
+        drainer: Watch, //when signaled, will stop driving all conns in the pool, effectively draining the pool.
+    ) -> WorkloadHBONEPool {
+        let (tx, rx) = watch::channel(false);
+        debug!(
+            "constructing pool with {:#?} streams per conn",
+            cfg.pool_max_streams_per_conn
+        );
         Self {
-            pool: HyperPool::new(
-                hyper_util::client::legacy::pool::Config {
-                    idle_timeout: Some(Duration::from_secs(90)),
-                    max_idle_per_host: std::usize::MAX,
-                },
-                TokioExec,
-                Some(TokioTimer::new()),
-            ),
+            pool_notifier: Arc::new(tx),
+            pool_watcher: rx,
+            max_streamcount: cfg.pool_max_streams_per_conn,
+            // the number here is simply the number of unique src/dest keys
+            // the pool is expected to track before the inner hashmap resizes.
+            connected_pool: Arc::new(pingora_pool::ConnectionPool::new(50000)),
+            cfg,
+            socket_factory,
+            cert_manager,
+            drainer,
+            established_conn_writelock: Arc::new(RwLock::new(HashMap::new())),
         }
     }
-}
 
-#[derive(Clone)]
-pub struct TokioExec;
+    // Obtain a pooled connection. Will prefer to retrieve an existing conn from the pool, but
+    // if none exist, or the existing conn is maxed out on streamcount, will spawn a new one,
+    // even if it is to the same dest+port.
+    //
+    // If many `connects` request a connection  to the same dest at once, all will wait until exactly
+    // one connection is created, before deciding if they should create more or just use that one.
+    pub async fn connect(&mut self, key: WorkloadKey) -> Result<Client, Error> {
+        // TODO BML this may not be collision resistant/slow. It should be resistant enough for workloads tho.
+        let mut s = DefaultHasher::new();
+        key.hash(&mut s);
+        let hash_key = s.finish();
+        let pool_key = pingora_pool::ConnectionMeta::new(
+            hash_key,
+            GLOBAL_CONN_COUNT.fetch_add(1, Ordering::Relaxed),
+        );
```

Comment:
According to ConnectionMeta documentation every connection that shares this hash_key is considered the same for the purposes of connection reuse. This suggests that hash collision could be pretty problematic, potentially even a security vulnerability, yeah? I'm not sure we can TODO something like that. It seems like we need an implementation which is better than just collision resistant sooner rather than later.

Reply:
It's a TODO for me - hash functions grow on trees, so I'm not very worried about dropping in a better one.

Reply:
I'm going to look at making Pingora's pool take an `impl Hash, Eq` but for now I decided to stick with this hasher and just bind ConnClients to workload keys.

Then do a final check before returning to prove that the conn we're giving you belongs to the key you gave us.

This should be strictly safe, and I think the odds of collision are very low for us here, the workload would likely crash the box long before it would open enough conns to create a kind of collision.